### PR TITLE
Refactor: `cat`

### PR DESCRIPTION
**Pull request by trym-b**
Best reviewed commit by commit

### CODE REVIEW

File: cmd/cat/cat.go [hunk 1]
```diff
@@ -11,126 +10,65 @@ import (
...
-	offset             int64
-	recordNum          int
-	recordCount        int
-	fileName           string
-	filter             *filter.Filter
-	showWarcHeader     bool
-	showProtocolHeader bool
-	showPayload        bool
+type catConfig struct {
+	Offset             int64
```
Comment (maeb) [inline cmd/cat/cat.go:16]: Since the struct is only used in the `cat` package the fields need not be public.

File: cmd/cat/cat.go [hunk 2]
```diff
@@ -11,126 +10,65 @@ import (
...
-
-func runE(c *conf) error {
-	readFile(c, c.fileName)
-	return nil
-}
-
-func readFile(c *conf, fileName string) {
-	wf, err := gowarc.NewWarcFileReader(fileName, c.offset, gowarc.WithBufferTmpDir(viper.GetString(flag.TmpDir)))
-	defer func() { _ = wf.Close() }()
+func listRecords(config *catConfig, fileName string) ([]gowarc.WarcRecord, error) {
```
Comment (maeb) [inline cmd/cat/cat.go:26]: I understand changing the signature of the function would make it testable but it would potentially blow up the memory consumption when cat'ing one or more large WARC files.

I would suggest using a send channel as a parameter 

```golang
func listRecords(config *catConfig, fileName string, records chan<- string) error {
  defer close(records)
  warcFileReader, err := gowarc.NewWarcFileReader(fileName, config.Offset, gowarc.WithBufferTmpDir(viper.GetString(flag.TmpDir)))
  defer func() { _ = warcFileReader.Close() }()
  ...
  go func() {
    warcRecord, _, _, err := warcFileReader.Next()
    ...
    records <- warcRecord
    ...
    // if any errors during iteration we print them
   fmt.Fprintf(os.Stderr, ..:)
    ...
 }()
 return nil
}
```
then the caller can decide the buffering and perform the actual output:

```golang

bufSize := 100 // we can live with eg. 100 records in memory
records := make(chan gowarc.WarcRecord, bufSize?)

err := listRecords(config, filename, records)
if err != nil {
   // handle error from gowerc.NewWarcFileReader
}

// iterate over the record from the records channel
for record := range records {
  // print the record according to config
  ...
   // and a test can append it to a slice here 
}
```

File: cmd/cat/cat.go [hunk 3]
```diff
@@ -11,126 +10,65 @@ import (
...
+	warcFileReader, err := gowarc.NewWarcFileReader(fileName, config.Offset, gowarc.WithBufferTmpDir(viper.GetString(flag.TmpDir)))
```
Comment (maeb) [inline cmd/cat/cat.go:27]: Maybe add `tmpDir` to the config struct so we don't need to import viper in the `cat.go` file.

**Review by maeb — CHANGES_REQUESTED** [review]
Only `NewCommand` sh

File: cmd/cat/cat.go (hunk 1 above)
Comment (trym-b) [reply cmd/cat/cat.go:16]: Makes sense, will fix.

File: cmd/cat/cat.go (hunk 2 above)
Comment (trym-b) [reply cmd/cat/cat.go:26]: I see your point, but I would also argue that trying to ensure that this project runs with a specific performance requirements, without actually having any tests for those requirements is a recipe for failure. If we are to enforce specific memory footprint requirements without any testing what-so-ever, how do we even know if we fulfill those requirements today?

For example, what is a "large WARC file"? Kilobytes sized ones? GB? TB?
If this project is only ever going to be a CLI tool, does it matter that it uses more memory that it should?

Anyway, since memory consumption is, apparently, important for this tool, I quickly profiled the code.
The important bits, before:
image

After:
image

So you are correct that the memory footprint is different, and because of its importance I will remove that latest commit so the other two commits are no longer blocked by this PR.

The TL:DR is that it is difficult to enforce a certain memory footprint without benchmark test or automatic profiling.

File: cmd/cat/cat.go [hunk 4]
```diff
@@ -11,126 +10,65 @@ import (
...
-type conf struct {
-	offset             int64
-	recordNum          int
-	recordCount        int
-	fileName           string
-	filter             *filter.Filter
-	showWarcHeader     bool
-	showProtocolHeader bool
-	showPayload        bool
+type catConfig struct {
```
[the remark this answers is not in the document]
Comment (trym-b) [reply cmd/cat/cat.go]: Fixed

File: cmd/cat/cat.go (hunk 3 above)
Comment (trym-b) [reply cmd/cat/cat.go]: I think this is out of scope of this PR, as it now purely does refactoring

**Review by maeb — APPROVED** [review]
